### PR TITLE
[WIP]Feature/issue 660

### LINKED ISSUE #660: add numbers with underscore
ruby had this nice feature to add underscore between digits for numbers `1_000_000_000`
it would be great if goby can do it too.

### PR DESCRIPTION
**Pull request by shes50103**
- Implement underscore feature
- make sure underscore using in the right way

`1_2_3` => 123
`1_2.3_4` => 12.34
`1__2` => Error
`_1` => Error
`1_` => Error 

This closes #660

### CODE REVIEW

File: compiler/lexer/lexer.go [hunk 1]
```diff
@@ -264,12 +264,29 @@ func (l *Lexer) resetNosymbol() {
...
+}
+
 func (l *Lexer) readNumber() []rune {
-	position := l.position
-	for isDigit(l.ch) {
+	var input []rune
+
+	for isDigit(l.ch) || l.ch == '_' {
+
+		if isDigit(l.ch) {
```
Comment (SD10) [inline compiler/lexer/lexer.go]: IMO I think the lexer should preserve the underscores in the token. In terms of responsibility, I don't think a lexer should be responsible for any code transformation. Removing them at this stage seems sort of magical. 

IMO the token should be simplified at a later stage, maybe during parsing or eval?

**Comment by st0012** [comment]
Well I think @SD10 has a good point. Keeping those information to parser allows us to provide more sophisticated error message

**Comment by SD10** [comment]
Here we are already transforming the token into a literal for `Int`:
https://github.com/goby-lang/goby/blob/c00ad0ca3d6b779b6ab737dea906637ac62e4708/compiler/parser/data_type_parsing.go#L12
Here for Float:
https://github.com/goby-lang/goby/blob/c00ad0ca3d6b779b6ab737dea906637ac62e4708/compiler/parser/data_type_parsing.go#L26

This is just my opinion. I'm a newbie 😊

**Comment by SD10** [comment]
It's also worth noting that in the Ruby lexer the underscores are maintained:
```
...
[[2, 7], :on_int, "1_000_000"],
```

**Pull request closed without merging**